sbncode/CAFMaker/CAFMaker_module.cc

```diff
@@ -1010,10 +1013,36 @@ void CAFMaker::produce(art::Event& evt) noexcept {
   GetByLabelStrict(evt, fParams.CRTHitLabel(), crthits_handle);
   // fill into event
   if (crthits_handle.isValid()) {
+
+    //==== gate start time
+    long long m_gate_start_timestamp = 0;
+    if(isRealData){
+
+      art::Handle< std::vector<raw::ExternalTrigger> > externalTrigger_handle;
+      evt.getByLabel( fParams.TriggerLabel(), externalTrigger_handle );
+      const std::vector<raw::ExternalTrigger> &externalTrgs = *externalTrigger_handle;
+
+      art::Handle< std::vector<raw::Trigger> > trigger_handle;
+      evt.getByLabel( fParams.TriggerLabel(), trigger_handle );
+      const std::vector<raw::Trigger> &trgs = *trigger_handle;
+
+      if(externalTrgs.size()==1 && trgs.size()==1){
+        long long TriggerAbsoluteTime = externalTrgs.at(0).GetTrigTime(); // Absolute time of trigger
+        double BeamGateRelativeTime = trgs.at(0).BeamGateTime(); // BeamGate time w.r.t. electronics clock T0 in us
+        double TriggerRelativeTime = trgs.at(0).TriggerTime(); // Trigger time w.r.t. electronics clock T0 in us
+        m_gate_start_timestamp = TriggerAbsoluteTime + (int)(BeamGateRelativeTime*1000-TriggerRelativeTime*1000);
+      }
+      else{
+        std::cerr << "Size of raw::ExternalTrigger = " << externalTrgs.size() << std::endl;
+        std::cerr << "Size of raw::Trigger = " << trgs.size() << std::endl;
+      }
+
+    }
```

Comment:
Sorry, tagged the wrong parenthesis. This comment, and the previous one, are trying to suggest code like this
```suggestion
    }
    else{ // MC
      // Big comment here
      m_gate_start_timestamp = 1600000;
    }
```
which means you can remove the MC case from FillCRTHit and treat everything equivalently.

Reply:
I updated this in the new commit